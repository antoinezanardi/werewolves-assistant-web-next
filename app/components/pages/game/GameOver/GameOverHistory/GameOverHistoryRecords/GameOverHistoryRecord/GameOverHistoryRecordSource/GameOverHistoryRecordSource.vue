<template>
  <div
    id="game-over-history-record-source"
    class="flex flex-col gap-4 item-center justify-center text-center"
  >
    <h4 id="game-over-history-record-source-name">
      {{ gameHistoryRecordPlaySourceNameLabel }}
    </h4>

    <PlayersHorizontalList
      id="game-over-history-record-source-horizontal-list"
      :players="props.gameHistoryRecord.play.source.players"
      :role-image-sizes="70"
    />
  </div>
</template>

<script setup lang="ts">
import type { GameOverHistoryRecordSourceProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordSource/game-over-history-record-source.types";
import PlayersHorizontalList from "~/components/shared/game/player/PlayersHorizontalList/PlayersHorizontalList.vue";
import { useGameSourceName } from "~/composables/api/game/useGameSource";

const props = defineProps<GameOverHistoryRecordSourceProps>();

const { getDefiniteGameSourceNameLabel } = useGameSourceName();

const gameHistoryRecordPlaySourceNameLabel = computed<string>(() => {
  const { source } = props.gameHistoryRecord.play;

  return getDefiniteGameSourceNameLabel(source.name, source.players.length);
});
</script>