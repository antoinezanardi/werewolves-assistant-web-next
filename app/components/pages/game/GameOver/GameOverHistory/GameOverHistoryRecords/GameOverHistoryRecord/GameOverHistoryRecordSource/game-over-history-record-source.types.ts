import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";

type GameOverHistoryRecordSourceProps = {
  gameHistoryRecord: GameHistoryRecord;
};

export type { GameOverHistoryRecordSourceProps };