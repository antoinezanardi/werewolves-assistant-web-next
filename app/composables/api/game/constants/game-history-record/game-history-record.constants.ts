const GAME_HISTORY_RECORD_PLAY_VOTING_RESULTS = [
  "sheriff-election",
  "tie",
  "death",
  "inconsequential",
  "skipped",
] as const;

const GAME_HISTORY_RECORD_PLAYER_ATTRIBUTE_ALTERATION_STATUSES = [
  "attached",
  "detached",
  "activated",
] as const;

export {
  GAME_HISTORY_RECORD_PLAY_VOTING_RESULTS,
  GAME_HISTORY_RECORD_PLAYER_ATTRIBUTE_ALTERATION_STATUSES,
};