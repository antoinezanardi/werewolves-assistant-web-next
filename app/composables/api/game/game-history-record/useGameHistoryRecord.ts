import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";

type UseGameHistoryRecord = {
  didSourceSkipped: (gameHistoryRecord: GameHistoryRecord) => boolean;
};

function useGameHistoryRecord(): UseGameHistoryRecord {
  function didSourceSkipped(gameHistoryRecord: GameHistoryRecord): boolean {
    const {
      type,
      votes,
      targets,
      didJudgeRequestAnotherVote,
      chosenCard,
    } = gameHistoryRecord.play;

    return type === "vote" && !votes ||
      type === "target" && !targets ||
      type === "request-another-vote" && didJudgeRequestAnotherVote === undefined ||
      type === "choose-card" && !chosenCard;
  }
  return { didSourceSkipped };
}

export { useGameHistoryRecord };