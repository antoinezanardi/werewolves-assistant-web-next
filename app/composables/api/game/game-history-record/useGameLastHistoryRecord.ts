import type { MaybeRef, Ref } from "vue";
import type { GameHistoryRecordPlayerAttributeAlterationStatus } from "~/composables/api/game/types/game-history-record/game-history-record-player-attribute-alteration/game-history-record-player-attribute-alteration.types";
import type { GameHistoryRecord } from "~/composables/api/game/types/game-history-record/game-history-record.class";
import type { Game } from "~/composables/api/game/types/game.class";
import type { GameSource } from "~/composables/api/game/types/game.types";
import type { PlayerAttributeName } from "~/composables/api/game/types/players/player-attribute/player-attribute.types";
import type { Player } from "~/composables/api/game/types/players/player.class";

type UseGameLastHistoryRecord = {
  lastTargetedPlayers: Ref<Player[]>;
  doesHavePlayerAttributeAlteration: (attributeName: PlayerAttributeName, source: GameSource, status: GameHistoryRecordPlayerAttributeAlterationStatus) => boolean;
};

function useGameLastHistoryRecord(game: MaybeRef<Game>): UseGameLastHistoryRecord {
  const lastGameHistoryRecord = computed<GameHistoryRecord | null>(() => (isRef(game) ? game.value.lastGameHistoryRecord : game.lastGameHistoryRecord));

  const lastTargetedPlayers = computed<Player[]>(() => {
    if (!lastGameHistoryRecord.value?.play.targets) {
      return [];
    }
    return lastGameHistoryRecord.value.play.targets.map(target => target.player);
  });

  function doesHavePlayerAttributeAlteration(attributeName: PlayerAttributeName, source: GameSource, status: GameHistoryRecordPlayerAttributeAlterationStatus): boolean {
    if (!lastGameHistoryRecord.value?.playerAttributeAlterations) {
      return false;
    }
    return lastGameHistoryRecord.value.playerAttributeAlterations.some(playerAttributeAlteration => playerAttributeAlteration.name === attributeName &&
      playerAttributeAlteration.source === source &&
      playerAttributeAlteration.status === status);
  }
  return {
    lastTargetedPlayers,
    doesHavePlayerAttributeAlteration,
  };
}

export { useGameLastHistoryRecord };