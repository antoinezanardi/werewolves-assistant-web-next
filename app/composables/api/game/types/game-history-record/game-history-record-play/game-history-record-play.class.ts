import { Expose, plainToInstance, Type } from "class-transformer";

import { GameAdditionalCard } from "~/composables/api/game/types/game-additional-card/game-additional-card.class";
import { GameHistoryRecordPlaySource } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-source/game-history-record-play-source.class";
import { GameHistoryRecordPlayTarget } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-target/game-history-record-play-target.class";
import { GameHistoryRecordPlayVote } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-vote/game-history-record-play-vote.class";
import { GameHistoryRecordPlayVoting } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play-voting/game-history-record-play-voting.class";
import type { GamePlayAction, GamePlayCause, GamePlayType } from "~/composables/api/game/types/game-play/game-play.types";
import type { RoleSide } from "~/composables/api/role/types/role.types";
import { DEFAULT_PLAIN_TO_INSTANCE_OPTIONS } from "~/utils/constants/class-transformer.constants";

class GameHistoryRecordPlay {
  @Expose()
  public type: GamePlayType;

  @Expose()
  public action: GamePlayAction;

  @Expose()
  @Type(() => GameHistoryRecordPlaySource)
  public source: GameHistoryRecordPlaySource;

  @Expose()
  public cause?: GamePlayCause;

  @Expose()
  @Type(() => GameHistoryRecordPlayTarget)
  public targets?: GameHistoryRecordPlayTarget[];

  @Expose()
  @Type(() => GameHistoryRecordPlayVote)
  public votes?: GameHistoryRecordPlayVote[];

  @Expose()
  @Type(() => GameHistoryRecordPlayVoting)
  public voting?: GameHistoryRecordPlayVoting;

  @Expose()
  public didJudgeRequestAnotherVote?: boolean;

  @Expose()
  @Type(() => GameAdditionalCard)
  public chosenCard?: GameAdditionalCard;

  @Expose()
  public chosenSide?: RoleSide;

  public static create(gameHistoryRecordPlay: GameHistoryRecordPlay): GameHistoryRecordPlay {
    return plainToInstance(GameHistoryRecordPlay, gameHistoryRecordPlay, DEFAULT_PLAIN_TO_INSTANCE_OPTIONS);
  }
}

export { GameHistoryRecordPlay };