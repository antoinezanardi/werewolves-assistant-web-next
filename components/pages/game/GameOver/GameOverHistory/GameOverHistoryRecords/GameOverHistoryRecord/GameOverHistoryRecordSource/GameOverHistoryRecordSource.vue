<template>
  <div
    id="game-over-history-record-source"
    class="flex flex-col gap-4 item-center justify-center text-center"
  >
    <h4>
      {{ gameHistoryRecordPlaySourceNameLabel }}
    </h4>

    <div
      id="game-over-history-record-source-players"
      class="flex gap-4 items-center justify-center"
    >
      <div
        v-for="playerSource in shuffledTruncatedSourcePlayers"
        :key="playerSource._id"
        class="game-over-history-record-source-player w-1/4"
      >
        <PlayerCard
          :player-name="playerSource.name"
          :player-role="playerSource.role.current"
        />
      </div>

      <VuePrimeTag
        v-if="areSourcePlayersTruncated"
        class="game-over-history-record-source-players-truncated w-1/4"
        severity="info"
        :value="truncatedSourcePlayersCountText"
      />
    </div>
  </div>
</template>

<script setup lang="ts">
import { shuffle } from "radash";
import type { GameOverHistoryRecordSourceProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordSource/game-over-history-record-source.types";
import PlayerCard from "~/components/shared/game/player/PlayerCard/PlayerCard.vue";
import type { Player } from "~/composables/api/game/types/players/player.class";
import { useGameSourceName } from "~/composables/api/game/useGameSource";

const props = defineProps<GameOverHistoryRecordSourceProps>();

const { getDefiniteGameSourceNameLabel } = useGameSourceName();

const gameHistoryRecordPlaySourceNameLabel = computed<string>(() => {
  const { source } = props.gameHistoryRecord.play;

  return getDefiniteGameSourceNameLabel(source.name, source.players.length);
});

const maximumSourcePlayersDisplayed = 3;

const areSourcePlayersTruncated = computed<boolean>(() => props.gameHistoryRecord.play.source.players.length > maximumSourcePlayersDisplayed);

const shuffledTruncatedSourcePlayers = computed<Player[]>(() => shuffle(props.gameHistoryRecord.play.source.players).slice(0, maximumSourcePlayersDisplayed));

const truncatedSourcePlayersCount = computed<number>(() => props.gameHistoryRecord.play.source.players.length - maximumSourcePlayersDisplayed);

const truncatedSourcePlayersCountText = computed<string>(() => `+ ${truncatedSourcePlayersCount.value}`);
</script>