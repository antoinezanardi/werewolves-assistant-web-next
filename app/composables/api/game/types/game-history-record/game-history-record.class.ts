import { Expose, plainToInstance, Type } from "class-transformer";

import { GameHistoryRecordPlay } from "~/composables/api/game/types/game-history-record/game-history-record-play/game-history-record-play.class";
import { GameHistoryRecordPlayerAttributeAlteration } from "~/composables/api/game/types/game-history-record/game-history-record-player-attribute-alteration/game-history-record-player-attribute-alteration.class";
import { GamePhase } from "~/composables/api/game/types/game-phase/game-phase.class";
import { Player } from "~/composables/api/game/types/players/player.class";
import { DEFAULT_PLAIN_TO_INSTANCE_OPTIONS } from "~/utils/constants/class-transformer.constants";

class GameHistoryRecord {
  @Expose()
  public _id: string;

  @Expose()
  public gameId: string;

  @Expose()
  public turn: number;

  @Type(() => GamePhase)
  @Expose()
  public phase: GamePhase;

  @Expose()
  public tick: number;

  @Expose()
  @Type(() => GameHistoryRecordPlay)
  public play: GameHistoryRecordPlay;

  @Expose()
  @Type(() => Player)
  public revealedPlayers?: Player[];

  @Expose()
  @Type(() => Player)
  public deadPlayers?: Player[];

  @Expose()
  @Type(() => GameHistoryRecordPlayerAttributeAlteration)
  public playerAttributeAlterations?: GameHistoryRecordPlayerAttributeAlteration[];

  @Expose()
  public createdAt: Date;

  public static create(gameHistoryRecord: GameHistoryRecord): GameHistoryRecord {
    return plainToInstance(GameHistoryRecord, gameHistoryRecord, DEFAULT_PLAIN_TO_INSTANCE_OPTIONS);
  }
}

export { GameHistoryRecord };