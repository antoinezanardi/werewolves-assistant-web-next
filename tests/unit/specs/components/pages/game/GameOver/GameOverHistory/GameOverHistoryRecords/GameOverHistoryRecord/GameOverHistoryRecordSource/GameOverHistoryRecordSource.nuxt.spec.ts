import { createFakeGameHistoryRecordPlaySource } from "@tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record-play/game-history-record-play-source/game-history-record-play-source.factory";
import { createFakeGameHistoryRecordPlay } from "@tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record-play/game-history-record-play.factory";
import { createFakeGameHistoryRecord } from "@tests/unit/utils/factories/composables/api/game/game-history-record/game-history-record.factory";
import { createFakeActorAlivePlayer } from "@tests/unit/utils/factories/composables/api/game/player/player-with-role.factory";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import type { GameOverHistoryRecordSourceProps } from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordSource/game-over-history-record-source.types";
import GameOverHistoryRecordSource from "~/components/pages/game/GameOver/GameOverHistory/GameOverHistoryRecords/GameOverHistoryRecord/GameOverHistoryRecordSource/GameOverHistoryRecordSource.vue";
import type PlayersHorizontalList from "~/components/shared/game/player/PlayersHorizontalList/PlayersHorizontalList.vue";
import type { Player } from "~/composables/api/game/types/players/player.class";

describe("Game Over History Record Source Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameOverHistoryRecordSource>>;
  const defaultSourcePlayers = [
    createFakeActorAlivePlayer({ name: "Antoine" }),
    createFakeActorAlivePlayer({ name: "Olivia" }),
    createFakeActorAlivePlayer({ name: "Thomas" }),
    createFakeActorAlivePlayer({ name: "Doudou" }),
  ];
  const defaultGameHistoryRecord = createFakeGameHistoryRecord({
    play: createFakeGameHistoryRecordPlay({
      type: "vote",
      action: "vote",
      source: createFakeGameHistoryRecordPlaySource({
        name: "survivors",
        players: defaultSourcePlayers,
      }),
    }),
  });
  const defaultProps: GameOverHistoryRecordSourceProps = { gameHistoryRecord: defaultGameHistoryRecord };

  async function mountGameOverHistoryRecordSourceComponent(options: ComponentMountingOptions<typeof GameOverHistoryRecordSource> = {}):
  Promise<ReturnType<typeof mount<typeof GameOverHistoryRecordSource>>> {
    return mountSuspendedComponent(GameOverHistoryRecordSource, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameOverHistoryRecordSourceComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Source Name", () => {
    it("should display source name when rendered.", () => {
      const sourceName = wrapper.find<HTMLHeadElement>("#game-over-history-record-source-name");

      expect(sourceName.text()).toBe("shared.game.player.group.definiteName.survivors, 4");
    });
  });

  describe("Source Players", () => {
    it("should pass source players to horizontal list when rendered.", () => {
      const horizontalList = wrapper.findComponent<typeof PlayersHorizontalList>("#game-over-history-record-source-horizontal-list");

      expect(horizontalList.props("players")).toStrictEqual<Player[]>(defaultSourcePlayers);
    });
  });
});