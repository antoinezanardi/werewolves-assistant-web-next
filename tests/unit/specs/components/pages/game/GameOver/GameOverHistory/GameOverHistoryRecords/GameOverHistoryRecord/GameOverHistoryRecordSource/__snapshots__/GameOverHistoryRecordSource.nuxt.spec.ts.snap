// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Over History Record Source Component > should match snapshot when rendered. 1`] = `
"<div id="game-over-history-record-source" class="flex flex-col gap-4 item-center justify-center text-center">
  <h4 id="game-over-history-record-source-name">shared.game.player.group.definiteName.survivors, 4</h4>
  <players-horizontal-list-stub players="[object Object],[object Object],[object Object],[object Object]" roleimagesizes="70" id="game-over-history-record-source-horizontal-list"></players-horizontal-list-stub>
</div>"
`;
